src/qibo/models/tsp.py

```diff
@@ -0,0 +1,160 @@
+                from QAOA
+
+                '''
+                small_tsp = TSP(distance_matrix)
+                obj_hamil, mixer = small_tsp.hamiltonians()
+                qaoa = QAOA(obj_hamil, mixer=mixer)
+                best_energy, final_parameters, extra = qaoa.minimize(initial_p=[0.1] * layer,
+                                                     initial_state=initial_state, method='BFGS')
+                qaoa.set_parameters(final_parameters)
+                quantum_state = qaoa.execute(initial_state)
+                meas = quantum_state.measure(gates.M(*range(9)), nshots=1000)
+                freq_counter = meas.frequencies()
+                # let's combine freq_counter here, first convert each key and sum up the frequency
+                cauchy_dict = defaultdict(int)
+                for freq_key in freq_counter:
+                    standard_cauchy_key = convert_to_standard_Cauchy(freq_key)
+                    cauchy_dict[standard_cauchy_key] += freq_counter[freq_key]
+                max_key = max(cauchy_dict, key=cauchy_dict.get)
+                return evaluate_dist(max_key)
+
+            np.random.seed(42)
+            num_cities = 3
+            distance_matrix = np.array([[0, 0.9, 0.8], [0.4, 0, 0.1],[0, 0.7, 0]])
+            distance_matrix = distance_matrix.round(1)
+            small_tsp = TSP(distance_matrix)
+            initial_parameters = np.random.uniform(0, 1, 2)
+            initial_state = small_tsp.prepare_initial_state([i for i in range(num_cities)])
+            qaoa_function_of_layer(2, distance_matrix)
+
+    """
+
+    def __init__(self, distance_matrix, backend=None):
+        if backend is None:
+            from qibo.backends import GlobalBackend
+            self.backend = GlobalBackend()
+        else:
+            self.backend = backend
+        self.distance_matrix = distance_matrix
+        self.num_cities = distance_matrix.shape[0]
+        self.two_to_one = calculate_two_to_one(self.num_cities)
+
+    def hamiltonians(self):
+        """
+        Returns:
+            The pair of Hamiltonian describes the phaser hamiltonian
+            and the mixer hamiltonian.
+
+        """
+        return (tsp_phaser(self.distance_matrix, backend=self.backend),
+                tsp_mixer(self.num_cities, backend=self.backend))
+
+    def prepare_initial_state(self, ordering):
+        """
+        To run QAOA by Hadsfield, we need to start from a valid permutation function to ensure feasibility.
+
+        Args:
+            ordering (array): A list describing permutation from 0 to n-1
+
+        Returns:
+            An initial state that is used to start TSP QAOA.
+
+        """
+        c = Circuit(len(ordering) ** 2)
+        for i in range(len(ordering)):
+            c.add(gates.X(int(self.two_to_one[ordering[i], i])))
+        result = self.backend.execute_circuit(c)
+        return result.state(numpy=True)
```

Comment:
I would suggest to drop numpy here:
```suggestion
        return result.state()
```